[CDAP-20922] Reuse the Spark ClassLoader in app-fabric

chtyim opened the pull request:
- For non local spark program execution purposes, we don’t need to isolate the Spark classloader, since there is no creation of the SparkContext object.

dli357 commented on cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/SparkProgramRuntimeProvider.java:
```diff
@@ -180,49 +178,68 @@ public boolean isSupported(ProgramType programType, CConfiguration cConf) {
+        // Never needs to rewrite yarn client in CDAP master, which is the only place using
+        // distributed program runner. Also wrap the classloader so that it can't be closed.
+        distributedRunnerClassLoader = new ClassLoader(createClassLoader(true, false, false)) {};
+        return distributedRunnerClassLoader;
```
Instead of returning `distributedRunnerClassLoader` and incurring another expensive read from a volatile variable, can we write the new classloader to our existing local variable `classLoader`, assign it to `distributedRunnerClassLoader`, then return the local variable?

chtyim replied:
That's not really expensive and it is just have it hit once due to the singleton. Just don't wanna clutter the code too much

chtyim committed:
[CDAP-20910] Reuse the Spark ClassLoader in app-fabric

- For non local spark program execution purposes, we don’t need to isolate the Spark classloader, since there is no creation of the SparkContext object.

arjan-bal commented on cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/SparkProgramRuntimeProvider.java:
```diff
@@ -180,49 +192,68 @@ public boolean isSupported(ProgramType programType, CConfiguration cConf) {
+      }
+      try {
+        // Never needs to rewrite yarn client in CDAP master, which is the only place using
+        // distributed program runner. Also wrap the classloader so that it can't be closed.
```
Wrapping it will also make it difficult for task workers to close the class loader. So we won't be able to solve the memory increases in task workers until we have a way to re-use this class loader in task workers also.
